### PR TITLE
Add tooling to generate OperatorHub release artifacts

### PR DESCRIPTION
**Pull request by nikhil-thomas**
# Changes

Add tooling to generate OperatorHub release artifacts

Modify resources in `config/bases` and `/config/openshift` to make it
work with operator bundle generation tools

Modify `kustomization.yaml`(s)

Add a csv template for generating release artifacts for Kubernetes

Add first set of generated bundle for kubernetes and openshift platforms

Future Work:

- write a Tekton Task (like TEP table checker) to ensure that latest version of bundles are  pushed with each PR.
- add this bundle generation tool as a Task in Operator release  pipeline.
-  configure bundle generation config.yaml for OpenShift

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [X] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -89,3 +89,21 @@ generated: | vendor ; $(info $(M) update generated files) ## Update generated fi
...
+		--default-channel stable \
+		--kustomize-dir csv-stub \
+		--overwrite \
+		--package openshift-pipelines-operator-rh \
+		--version 0.22.0-1
+	rm -rf csv-stub
+
+
+bundle:
+	$(MAKE) -C operatorhub/openshift --file ../../Makefile bundlegen
```
[the remark this answers is not in the document]
Comment (nikhil-thomas) [reply Makefile:109]: yup definitely.

File: Makefile [hunk 2]
```diff
@@ -46,7 +48,7 @@ $(BIN)/golangci-lint: | $(BIN) ; $(info $(M) getting golangci-lint $(GOLANGCI_VE
 
 .PHONY: clean-cluster
 clean-cluster: | $(KO) $(KUSTOMIZE) clean-cr; $(info $(M) clean $(TARGET)…) @ ## Cleanup cluster
-	-$(KUSTOMIZE) build config/$(TARGET) | $(KO) delete -f -
+	-$(KUSTOMIZE) build --load-restrictor LoadRestrictionsNone config/$(TARGET) | $(KO) delete -f -
```
Comment (vdemeester) [inline Makefile:51]: What does `--load-restrictor LoadRestrictionsNone` and why is it needed ?

File: config/base/kustomization.yaml [hunk 3]
```diff
@@ -17,16 +17,13 @@ resources:
 - 100-namespace.yaml
 - 300-operator_v1alpha1_pipeline_crd.yaml
 - 300-operator_v1alpha1_trigger_crd.yaml
-- 300-operator_v1alpha1_dashboard_crd.yaml
-- 300-operator_v1alpha1_addon_crd.yaml
 - 300-operator_v1alpha1_config_crd.yaml
-- 300-operator_v1alpha1_result_crd.yaml
```
Comment (vdemeester) [inline config/base/kustomization.yaml:23]: Why are those removed from here ?

File: config/openshift/operator.yaml [hunk 4]
```diff
@@ -15,19 +15,39 @@
 apiVersion: apps/v1
 kind: Deployment
 metadata:
-  name: tekton-operator
+  name: openshift-pipelines-operator
```
Comment (vdemeester) [inline config/openshift/operator.yaml:18]: Any reason for the name change here ? This is `tektoncd/operator` not the *downstream* openshift pipeline operator 👼🏼

File: operatorhub/Makefile [hunk 5]
```diff
@@ -0,0 +1,12 @@
+BUNDLE_ARGS := $(or ${BUNDLE_ARGS},--help)
+
+OPERATOR_SDK = $(or ${OPERATOR_SDK_BIN},$(BIN)/operator-sdk)
+
+$(BIN)/operator-sdk: | $(BIN) ; $(info $(M) getting operator-sdk)
+	ls $(BIN)
+	@./operatorhub/tools/install_operator-sdk.sh $(BIN)
```
Comment (vdemeester) [inline operatorhub/Makefile:7]: Is this path right ? 🤔

File: operatorhub/kubernetes/README.md [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+bundle for using `release-manifest` strategy.
+
+## Steps to generate bundles from an existing release.yaml
+
+**Note:** The input release.yaml could be a github release or the result of `ko resolve config`
+
+1. Make sure that we have the release manifest file ready
+   eg:
+   ```bash
+   curl -L -o tektoncd-operaotr-0.49.0 https://github.com/tektoncd/operator/releases/download/v0.49.0/release.notags.yaml
```
Comment (vdemeester) [inline operatorhub/kubernetes/README.md:13]: typo: tektoncd-operator-*

File: operatorhub/kubernetes/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    make operator-bundle
+    ```
+
+    **CLI flags explained**
+
+    Flag                                          | Description
+    --------------------------------------------- | -----------
+    `--workspace operatorhub/kubernetes`          | the working directory where the operator bundle should be assembled
+    `--operator-release-version 0.49.0`           | version of the release (version of bundle)
+    `--channels alpha`                            | target release channel(s) (eg: stabe,preview)
```
Comment (vdemeester) [inline operatorhub/kubernetes/README.md:29]: typo: stable

File: operatorhub/kubernetes/config.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+platform: "kubernetes"
+operator-packagename: "tektoncd-operator"
+image-substitutions:
+- image: ko://github.com/tektoncd/operator/cmd/kubernetes/operator:develContainer
```
Comment (vdemeester) [inline operatorhub/kubernetes/config.yaml:4]: What `develContainer` means ?

File: operatorhub/kubernetes/config.yaml [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  replaceLocations:
+    containerTargets:
+    - deploymentName: tekton-operator
+      containerName: tekton-operator
+- image: ko://github.com/tektoncd/operator/cmd/kubernetes/webhook:develEnv
```
Comment (vdemeester) [inline operatorhub/kubernetes/config.yaml:9]: Same for `develEnv`

File: operatorhub/kubernetes/manifests/bases/tektoncd-operator.clusterserviceversion.template.yaml [hunk 10]
```diff
@@ -0,0 +1,218 @@
...
+      name: tektonpipelines.operator.tekton.dev
+      version: v1alpha1
+    - description: |
+        TektonTrigger installs Tekton Triggers on a cluster. The operator installs Tekton Triggers on a cluster
+        if an instnce of this CRD is created on your cluster with name `trigger`. At present, the operator honors only
+        one CustomResource instance with the name `trigger` to ensure only one set of Tekton Triggers exists on a cluster.
+      displayName: TektonTriggers
+      kind: TektonTrigger
+      name: tektontriggers.operator.tekton.dev
+      version: v1alpha1
```
Comment (vdemeester) [inline operatorhub/kubernetes/manifests/bases/tektoncd-operator.clusterserviceversion.template.yaml:52]: This part is dynamic isn't it ? When we add(ed) `TektonResult`, `TektonChain`, … this will need to evolve

File: operatorhub/kubernetes/release-artifacts/bundle/manifests/config-logging_v1_configmap.yaml [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        "timeEncoder": "",
+        "durationEncoder": "",
+        "callerEncoder": ""
+      }
+    }
+kind: ConfigMap
+metadata:
+  labels:
+    operator.tekton.dev/release: devel
+  name: config-logging
```
Comment (vdemeester) [inline operatorhub/kubernetes/release-artifacts/bundle/manifests/config-logging_v1_configmap.yaml:72]: Should we, can we, name this differently ? (prefixing with `tekton` for example)

**Review by vdemeester — COMMENTED** [review]
- What do we gain by duplicating webhook and operator deployment (between kubernetes and openshift platform) ? 
- Are files in `release-artifacts/bundle/{manifests, …}` supposed to be in the git index ?

File: Makefile (hunk 2 above)
Comment (nikhil-thomas) [reply Makefile:51]: in latest version of `kustomize` we cannot bring in resources which are outside base root.
In our case, the kustomization for bundle generation is defined in `operatorhub/<platform>/kustomization.yaml` . This pulls resources from:

- `operatorhub/<platform>/base`
- `operatorhub/<platform>/overlay/<strategy-name>`
to get csv stub and
- `../../config/<platform-name>` (to get all necessary resources)

the `../../config/` will not work with kustomize if we don't have that flag.

Comment (nikhil-thomas) [reply Makefile:51]: The need for all this is to make the resource accumulation for CSV generation as close to how it is done in an Operator-sdk project.

eg: 
- input collection is specified through this: https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/memcached-operator/config/manifests
- so that the bundle will be generated here:https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/memcached-operator/bundle

File: config/base/kustomization.yaml (hunk 3 above)
Comment (nikhil-thomas) [reply config/base/kustomization.yaml:23]: As dashboard, results and addons are not common in all platforms, we need to move it to respective `<overlay-name>/kustomization.yaml` instead of keeping them in `base/kustomization.yaml`.

File: Makefile (hunk 2 above)
Comment (vdemeester) [reply Makefile:51]: I see. Can we have a small comment about this ? 😛

File: config/base/kustomization.yaml (hunk 3 above)
Comment (vdemeester) [reply config/base/kustomization.yaml:23]: Ok, then the question is "why are they not common ?" 😝 `Results` for example, or others.

File: config/openshift/operator.yaml (hunk 4 above)
Comment (vdemeester) [reply config/openshift/operator.yaml:18]: @nikhil-thomas I know it is 😝 but still, this is `tektoncd/operator`, not our *downstream* openshift-pipelines operator. This is the base of openshift-pipelines operator, but should it have the same name here, I am really not sure 🙃

File: operatorhub/Makefile (hunk 5 above)
Comment (vdemeester) [reply operatorhub/Makefile:7]: as.. we are in `operatorhub` folder already, so if I run the `Makefile` while being in `operatorhub` folder, it is gonna fail isn't it ?

File: Makefile (hunk 2 above)
Comment (nikhil-thomas) [reply Makefile:51]: sure 😝 👍

File: config/base/kustomization.yaml (hunk 3 above)
Comment (nikhil-thomas) [reply config/base/kustomization.yaml:23]: ie, TektonResults are not supported in OpenShift builds yet. :slightly_smiling_face:
so when we add the TektonResult controller to OpenSHift main, we shall make this common :thumbsup:

File: operatorhub/kubernetes/config.yaml (hunk 8 above)
Comment (nikhil-thomas) [reply operatorhub/kubernetes/config.yaml:4]: oops.. it was a debugging effort 😄 
removed 👍

File: operatorhub/kubernetes/config.yaml (hunk 9 above)
Comment (nikhil-thomas) [reply operatorhub/kubernetes/config.yaml:9]: typo

File: operatorhub/kubernetes/manifests/bases/tektoncd-operator.clusterserviceversion.template.yaml (hunk 10 above)
Comment (nikhil-thomas) [reply operatorhub/kubernetes/manifests/bases/tektoncd-operator.clusterserviceversion.template.yaml:52]: This file is a stub/seed from which CSV is built (not dynamic).

- Description text for CRDs
- icon
- links, keywords, etc

stuff that don't change between releases. 

But yes, TektonResult, TektonChain addition will need an entry in this file.

File: operatorhub/kubernetes/release-artifacts/bundle/manifests/config-logging_v1_configmap.yaml (hunk 11 above)
Comment (nikhil-thomas) [reply operatorhub/kubernetes/release-artifacts/bundle/manifests/config-logging_v1_configmap.yaml:72]: absolutely. 
in this bundle, it was generated from v0.49.0 release. that is why the naming is old.
when we generate bundle again the names will be automatically updated.

File: operatorhub/Makefile (hunk 5 above)
Comment (nikhil-thomas) [reply operatorhub/Makefile:7]: we are using `include operatorhub/Makefile` in the main Makefile. So we are not using a `make -c` in the call to `operator-bundle`. The working dir will be repository root.

File: config/crs/kubernetes/kustomization.yaml [hunk 12]
```diff
@@ -0,0 +1,8 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
```
Comment (sm43) [inline config/crs/kubernetes/kustomization.yaml:1]: how is this kustomization file used?
do we package cr examples?

File: config/openshift/webhook.yaml [hunk 13]
```diff
@@ -16,9 +16,37 @@ apiVersion: apps/v1
 kind: Deployment
 metadata:
   name: tekton-operator-webhook
```
Comment (sm43) [inline config/openshift/webhook.yaml:18]: as the controller deployment is renamed to `openshift-` do we waana rename this also?

File: operatorhub/kubernetes/release-artifacts/bundle/manifests/operator.tekton.dev_tektonconfigs.yaml [hunk 14]
```diff
@@ -0,0 +1,40 @@
+apiVersion: apiextensions.k8s.io/v1
```
Comment (sm43) [inline operatorhub/kubernetes/release-artifacts/bundle/manifests/operator.tekton.dev_tektonconfigs.yaml:1]: so we have all crds in `release-artifacts` are they manually added?
and do we need to keep them here or are they copied while creating csv?

**Comment by nikhil-thomas** [comment]
> * What do we gain by duplicating webhook and operator deployment (between kubernetes and openshift platform) ?
> * Are files in `release-artifacts/bundle/{manifests, …}` supposed to be in the git index ?

it is to deal with the difference in metadata, ENV etc. 
In ther future, I shall extract the common parts of spec into the base operator.yaml.

File: config/base/kustomization.yaml (hunk 3 above)
Comment (nikhil-thomas) [reply config/base/kustomization.yaml:23]: i have created an issue https://github.com/tektoncd/operator/issues/400
let us continue this discussion therer 🙂

File: config/openshift/operator.yaml (hunk 4 above)
Comment (nikhil-thomas) [reply config/openshift/operator.yaml:18]: i shall work on this as part of configuring the operator bundle generation for openshift.
shall we merge this pr as it is. 🙂

**Review by vdemeester — APPROVED** [review]

File: config/crs/kubernetes/kustomization.yaml (hunk 12 above)
Comment (nikhil-thomas) [reply config/crs/kubernetes/kustomization.yaml:1]: it is used [here](https://github.com/nikhil-thomas/tektoncd-operator/blob/3bb89e38cdf10aaafe2f138cac9b2c3607d71943/operatorhub/kubernetes/manifests/fetch-strategy-local/kustomization.yaml#L3)

the cr examples will be filled into the [csv stub/seed](https://github.com/nikhil-thomas/tektoncd-operator/blob/3bb89e38cdf10aaafe2f138cac9b2c3607d71943/operatorhub/kubernetes/manifests/bases/tektoncd-operator.clusterserviceversion.template.yaml#L5)

end result will look like [this](https://github.com/nikhil-thomas/tektoncd-operator/blob/3bb89e38cdf10aaafe2f138cac9b2c3607d71943/operatorhub/kubernetes/release-artifacts/bundle/manifests/tektoncd-operator.clusterserviceversion.yaml#L5)

File: config/openshift/webhook.yaml (hunk 13 above)
Comment (nikhil-thomas) [reply config/openshift/webhook.yaml:18]: we can do that.. Let me handle this in a follow up pr where i configure the config.yaml for openshift.

File: operatorhub/kubernetes/release-artifacts/bundle/manifests/operator.tekton.dev_tektonconfigs.yaml (hunk 14 above)
Comment (nikhil-thomas) [reply operatorhub/kubernetes/release-artifacts/bundle/manifests/operator.tekton.dev_tektonconfigs.yaml:1]: as part of the bundle generation, CRDs, configmaps etc in the accumulated list (from local or operator release.yaml) ends up as seprate files beside the ClusterSerciviceVersion file.

File: config/crs/kubernetes/kustomization.yaml (hunk 12 above)
Comment (sm43) [reply config/crs/kubernetes/kustomization.yaml:1]: got it. the cr examples doesn't have all newly added fields for customization. we  need to update the cr example?

**Comment by sm43** [comment]
lgtm
cc @savitaashture @pradeepitm12

File: config/crs/kubernetes/kustomization.yaml (hunk 12 above)
Comment (nikhil-thomas) [reply config/crs/kubernetes/kustomization.yaml:1]: we can, this pr is about adding the tooling. the Bundle is just there as an example.. 
the idea is that we need to regerate it with each commit. `make operator-bundle`.

i will add a ci-test (like TEP generated table test) to ensure that each pr has the updated bundles.

so examples, additional metadata can be regenerated at anypoint.

Comment (nikhil-thomas) [reply config/crs/kubernetes/kustomization.yaml:1]: and let us add the cr examples in a different patch.

File: test/resources/tektonconfigs.go [hunk 15]
```diff
@@ -50,7 +51,9 @@ func EnsureTektonConfigExists(kubeClientSet *kubernetes.Clientset, clients confi
 	}
 
 	tcCR, err := clients.Get(context.TODO(), names.TektonConfig, metav1.GetOptions{})
-
+	// this timeout is needed to make sure that the e2e tests work
+	// TODO: https://github.com/tektoncd/operator/issues/401
+	time.Sleep(120 * time.Second)
```
Comment (pradeepitm12) [inline test/resources/tektonconfigs.go:56]: by any chance is this going to affect the overall time of our test?

Comment (pradeepitm12) [reply test/resources/tektonconfigs.go:56]: Oh, my bad. somehow I took it as 2 hrs.

Comment (vdemeester) [reply test/resources/tektonconfigs.go:56]: We should use the `wait` package (or function) from `tektoncd/pipeline` here. `time.Sleep` is very very flakey prone 😓

Comment (nikhil-thomas) [reply test/resources/tektonconfigs.go:56]: practically no change was observed. 

This time out ensures that the test does not become flaky as sometimes, the tektonconfig instance will not be created by the time this line is executed. 

What we really need is a proper retry interval + timeout. (as described in the issue mentioned in the comment.)

Comment (vdemeester) [reply test/resources/tektonconfigs.go:56]: Yes, doing something similar to https://github.com/tektoncd/pipeline/blob/main/test/wait.go#L69 would help and make it even less flaky 😝

Comment (nikhil-thomas) [reply test/resources/tektonconfigs.go:56]: shall we handle it after https://github.com/tektoncd/operator/pull/385 merges. This changes how we define `ready for testing` state.

Comment (vdemeester) [reply test/resources/tektonconfigs.go:56]: yes of course, thanks for linking the comment in the issue.

**Review by pradeepitm12 — APPROVED** [review]
/lgtm
